ddtrace/profiling/collector/_memalloc_heap.c

```diff
@@ -351,7 +361,93 @@ memalloc_heap(void)
      * New allocations will go into the secondary freezer.allocs_m map and allocations
      * tracked in allocs_m which are freed will be added to a list to be removed when
      * the profiler is thawed. */
-    PyObject* heap_list = memalloc_heap_map_export(global_heap_tracker.allocs_m);
+
+    /* Calculate total number of samples: live + freed */
+    size_t live_count = memalloc_heap_map_size(global_heap_tracker.allocs_m);
+    size_t freed_count = global_heap_tracker.allocation_list.count;
+    size_t total_count = live_count + freed_count;
+
+    PyObject* heap_list = PyList_New(total_count);
+    if (heap_list == NULL) {
+        heap_tracker_thaw(&global_heap_tracker);
+        return NULL;
+    }
+
+    int list_index = 0;
+
+    /* First, iterate over live samples using the new iterator API */
+    memalloc_heap_map_iter_t* it = memalloc_heap_map_iter_new(global_heap_tracker.allocs_m);
+    if (it) {
+        void* key;
+        traceback_t* tb;
+
+        while (memalloc_heap_map_iter_next(it, &key, &tb)) {
+            if (list_index >= total_count) {
+                break;
+            }
+
+            PyObject* tb_and_info = PyTuple_New(4);
+            if (tb_and_info == NULL) {
+                continue;
+            }
+
+            size_t in_use_size = tb->size;
+            size_t alloc_size = tb->reported ? 0 : tb->size;
+
+            PyTuple_SET_ITEM(tb_and_info, 0, traceback_to_tuple(tb));
+            PyTuple_SET_ITEM(tb_and_info, 1, PyLong_FromSize_t(in_use_size));
+            PyTuple_SET_ITEM(tb_and_info, 2, PyLong_FromSize_t(alloc_size));
+            PyTuple_SET_ITEM(tb_and_info, 3, PyLong_FromSize_t(tb->count));
```

Comment:
I assume there is a reader somewhere which needs to be kept in sync with this object order?